packages/core/src/components/overlay2/overlay2.tsx

```diff
@@ -0,0 +1,710 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import classNames from "classnames";
+import * as React from "react";
+import { CSSTransition, TransitionGroup } from "react-transition-group";
+
+import { Classes } from "../../common";
+import {
+    OVERLAY_CHILD_REF_AND_REFS_MUTEX,
+    OVERLAY_CHILD_REQUIRES_KEY,
+    OVERLAY_WITH_MULTIPLE_CHILDREN_REQUIRES_CHILD_REFS,
+} from "../../common/errors";
+import { DISPLAYNAME_PREFIX, type HTMLDivProps } from "../../common/props";
+import {
+    ensureElement,
+    getActiveElement,
+    getRef,
+    isEmptyString,
+    isNodeEnv,
+    isReactElement,
+    setRef,
+    uniqueId,
+} from "../../common/utils";
+import { hasDOMEnvironment } from "../../common/utils/domUtils";
+import { usePrevious } from "../../hooks/usePrevious";
+import type { OverlayProps } from "../overlay/overlayProps";
+import { getKeyboardFocusableElements } from "../overlay/overlayUtils";
+import { Portal } from "../portal/portal";
+
+// HACKHACK: ugly global state manipulation should move to a proper React context
+const openStack: OverlayInstance[] = [];
+const getLastOpened = () => openStack[openStack.length - 1];
+
+/**
+ * HACKHACK: ugly global state manipulation should move to a proper React context
```

Comment:
consider fixing this in this PR and biting the bullet with the necessary `<OverlayProvider>` break